### PR TITLE
implement Channel, the base iodrivers_base::Driver class that uses protobufs

### PR DESCRIPTION
**Pull request by doudou**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,48 +1,94 @@
...
+protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS comms_base_station.proto)
 
-Rock CMake Macros
------------------
+rock_library(comms_base_station
+    USE_BINARY_DIR
+    HEADERS Device.hpp BaseStation.hpp ${PROTO_HDRS}
+    SOURCE ${PROTO_SRCS}
+    DEPS_PKGCONFIG comms_protobuf
+    DEPS_PLAIN Protobuf
```
Comment (g-arjones) [inline README.md:80]: Isn't this `DEPS_CMAKE`?

File: src/Channel.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                auto payload_range = protocol::getPayload(&m_receive_buffer[0]);
+
+                bool success = result.ParseFromString(
+                    std::string(reinterpret_cast<char const*>(payload_range.first),
+                                reinterpret_cast<char const*>(payload_range.second))
+                );
+                if (success) {
+                    return result;
+                }
+                else if (deadline < base::Time::now()) {
```
Comment (g-arjones) [inline src/Channel.hpp]: I'm not sure about silently ignoring validated packages that fails to be decoded like that

File: src/Protocol.cpp [hunk 3]
```diff
@@ -47,6 +48,28 @@ int protocol::extractPacket(uint8_t const* buffer, size_t size,
...
+
+uint8_t* protocol::encodeFrame(uint8_t* buffer,
+                               uint8_t const* payload_begin,
+                               uint8_t const* payload_end) {
+    buffer[0] = SYNC_0;
+    buffer[1] = SYNC_1;
+
+    size_t payload_length = payload_end - payload_begin;
+    uint8_t* length_end = encodeLength(buffer + 2, payload_length);
+    std::memcpy(length_end, payload_begin, payload_length);
```
Comment (g-arjones) [inline src/Protocol.cpp:75]: I think you should be checking the buffer size to make sure the payload fits here

File: src/Protocol.hpp [hunk 4]
```diff
@@ -33,16 +38,63 @@ namespace comms_protobuf {
...
+         */
+        uint8_t* encodeFrame(uint8_t* buffer,
+                             uint8_t const* payload_begin,
+                             uint8_t const* payload_end);
+
         /** Computes the 2-byte checksum over the given buffer
          */
         uint16_t crc(uint8_t const* begin, uint8_t const* end);
+
+        /** Encode a frame containing the given protobuf message */
```
Comment (g-arjones) [inline src/Protocol.hpp]: Should document the return value

File: README.md (hunk 1 above)
Comment (doudou) [reply README.md:80]: Since I've already done the `find_package`, it's just `DEPS_PLAIN`

File: src/Channel.hpp (hunk 2 above)
Comment (doudou) [reply src/Channel.hpp]: My take originally was that a valid low-level package that cannot be decoded by protobufs is functionally equivalent to an invalid package, i.e. this is the behavior that matches the iodrivers_base general behavior best.

Now, it's most probably a high-level error (i.e. incompatible messages). Drivers usually throw on those ... I'll move back to throwing in this case.

File: src/Protocol.cpp (hunk 3 above)
Comment (doudou) [reply src/Protocol.cpp:75]: Agreed. I went for the "the caller should know what he's doing", but that really is a micro-optimization with possibly catastrophic results. I'll check if there are other places where some sanity checks could be put.

File: src/Channel.hpp (hunk 2 above)
Comment (g-arjones) [reply src/Channel.hpp]: My point exactly.. You've already checked the crc in extractPacket... Any error past that point isn't really a comms error

**Comment by doudou** [comment]
I've made significant changes due to the more defensive coding style w.r.t. buffer length... Please re-review ;-)

File: manifest.xml [hunk 5]
```diff
@@ -6,9 +6,9 @@
 
   <depend package="base/cmake" />
   <depend package="drivers/iodrivers_base" />
-  <depend package="protobuf-compiler" />
   <depend package="protobuf-cxx" />
```
Comment (g-arjones) [inline manifest.xml]: This is a test_depend as well

File: src/Protocol.hpp [hunk 6]
```diff
@@ -33,16 +29,91 @@ namespace comms_protobuf {
...
+         *
+         * @return the actual encoded message length
+         */
+        size_t validateEncodingBufferSize(size_t buffer_length, size_t length);
+
+        /** Compute the size in bytes of the encoded version of the given length
+         */
+        size_t getLengthEncodedSize(size_t length);
+
+        /** Write variable-length integer field
```
Comment (g-arjones) [inline src/Protocol.hpp:64]: Documentation is outdated

File: manifest.xml (hunk 5 above)
Comment (doudou) [reply manifest.xml]: True ...

Comment (doudou) [reply manifest.xml]: Done

File: src/Protocol.hpp (hunk 6 above)
Comment (doudou) [reply src/Protocol.hpp:64]: Done

**Review by g-arjones — APPROVED** [review]

**Pull request merged**